### ISSUE TITLE
CAR geometries

### ISSUE THREAD

**Issue by AdriJD** [simonsobs/map_based_simulations#42]
We should have a standard CAR geometry to use for both the map-based simulations and the time-domain simulations. 

We can be guided by two points:

1. The analysis interface document (see [section 2.1.1](https://simons1.princeton.edu/docs/tod2maps_docs/analysis_interface/analysis_interface_LATEST.pdf)) specifies that we use a CAR geometry that spans roughly `-70 <= DEC <= 30` with X, Y resolution in deg/pix of `CDELT1 = −1/120` , `CDELT2 = 1/120`  and reference pixel index `CRPIX1 = 21601.0`, `CRPIX2 = 7561.0` at `RA, DEC = (0, 0)`, i.e. `CRVAL1 = 0.0, CRVAL2 = 0.0`. Note the negative sign of `CDELT1` indicating that RA decreases with increasing pixel index. 
2. The quadrature rules (integration weights) for the spherical harmonic transforms. Not all versions of CAR have a corresponding quadrature rule. Without a quadrature rule spherical harmonic analysis ("map2alm") becomes inaccurate. The geometry specified in point 1  will conform to the Clenshaw–Curtis quadrature rule, but will not correspond to a known rule when the maps are downgraded using `enmap.downgrade`. This is a very common step in many analysis pipelines. A more suitable geometry can be obtained by shifting the y pixel centers by half a pixel. This puts the pixel centers half a pixel width away from the north and south pole, which corresponds to the Fejer1 quadrature rule. This geometry stays a valid Fejer1 geometry when downgraded. Transforming a  Clenshaw–Curtis CAR map to a Fejer1 map corresponds to: `CRPIX2 -= 0.5`, `NAXIS2 -= 1`. Note that strictly speaking the second step is only necessary when the input map includes the north pole. Although I cannot find a written statement, I am quite sure we have decided that for SO we want to use the Fejer1 CAR variant. The ACT DR6 maps will likely also be also be released in the Fejer1 CAR variant.

Zach Atkins pointed out to me that the current `scan-s0003` TOD sims use a different CAR geometry. They use a positive `CDELT1` value and `CRVAL1 = 180.0`. This goes against the above specification where where RA=0 is in the middle of the map instead of at the edge. It also goes against the convention used within ACT and thus some of the plotting tools have some trouble with the positive `CDELT1` value (thanks to Zach for pointing this out). Reijo has made the geometry for these maps using [this script](https://github.com/simonsobs/pwg-scripts/blob/master/pwg-tds/scan-strategy-sims-2022/perlmutter/get_area_file.py) which is called using [this script]( https://github.com/simonsobs/pwg-scripts/blob/master/pwg-tds/scan-strategy-sims-2022/perlmutter/get_area_file.sh)

I propose we update the scripts used by Reijo such that it produces output that conforms to point 1, except for the half pixel shift that makes it conform to the Fejer1 rule.

The new script would be:
```
import argparse
import numpy as np
from pixell import utils, enmap, curvedsky

parser = argparse.ArgumentParser()
parser.add_argument("box", type=str)
parser.add_argument("ofile")
parser.add_argument("-r", "--res",  type=float, default=0.5)
parser.add_argument("-p", "--proj", type=str, default="car")

args = parser.parse_args()
box = utils.parse_box(args.box) * utils.degree
shape, wcs = enmap.geometry(box, res=args.res * utils.arcmin, proj=args.proj)

if wcs.wcs.cdelt[0] > 0:
    raise ValueError(f'By convention the X CDELT value should be negative, got : {wcs.wcs.cdelt[0]}')
if wcs.wcs.cdelt[1] < 0:
    raise ValueError(f'By convention the Y CDELT value should be positive, got : {wcs.wcs.cdelt[1]}')

# Check if geometry conforms to Clenshaw–Curtis. This will raise error if no quad rule can be found.
# This should never happen because the above always creates CC CAR maps.
curvedsky.get_minfo(shape, wcs, quad=True)

# Convert Clenshaw–Curtis CAR map to a Feyer1 CAR map.
wcs.wcs.crpix[1] -= 0.5

# Check if input map included the north pole, if so, we need to subtract one row.
# We do this by checking if the y pixel center of the top corner of the map has been shifted above +90 deg.
if np.degrees(enmap.corners(shape, wcs, corner=False))[1,0] > 90:
    shape = (shape[0] - 1, shape[1])

# Check if geometry conforms to Feyer1. 
curvedsky.get_minfo(shape, wcs, quad=True)

enmap.write_map(args.ofile, enmap.zeros(shape, wcs, np.uint8))
```

The above script can be called using an updated version of the calling script:
```
#!/bin/bash

# The coordinates are given in format Dec_min:Dec_max,RA_min:RA_max

python3 get_area_file.py --res 15  --proj car " -63:25,180:-180" area_file_15.0.fits
python3 get_area_file.py --res 5   --proj car " -63:25,180:-180" area_file_05.0.fits
python3 get_area_file.py --res 1.5 --proj car " -63:25,180:-180" area_file_01.5.fits
python3 get_area_file.py --res 0.5 --proj car " -63:25,180:-180" area_file_00.5.fits
```

The last line of the script: `python3 get_area_file.py --res 0.5 --proj car " -63:25,180:-180" area_file_00.5.fits` will now produce a geometry with:
```
CDELT1 = -0.00833333
CDELT2 = 0.00833333
CRPIX1 = 21601.0
CRPIX2 = 7560.5
CRVAL1 = 0.0
CRVAL2 = 0.0
NAXIS1 = 43200
NAXIS2 = 10560
```
Which matches the above requirements.

The current scan-s0003 maps can be modified in place I think. We just update the WCS. They are noise maps, so a half-pixel shift won't matter. No need to rerun those.

**Comment by zatkins2** [comment]
Thanks for summarizing this, Adri.

For modifying the curernt scan-s0003 maps, I agree your point about not needing to modify the data is valid in terms of moving to a Feyer1 quadrature rule, but for completeness, we need to make sure to modify the data to account for the change in CRVAL1 and CDELT1, namely, that column 0 of new map corresponds to column 21600 of old map, column 1 of new map corresponds to column 21599 of old map, etc. This can also happen on disk without need to rerun anything of course!

**Comment by AdriJD** [comment]
Yes I agree. Thanks for the note!

**Comment by zonca** [comment]
Thanks @AdriJD, could you make a pull request to `mapsims` that implements this?

The relevant files that need to be modified are listed in the pull request implementing the CAR output:
https://github.com/galsci/mapsims/pull/83/files

**Comment by zonca** [comment]
@AdriJD in `pysm3` here is how I handle CAR, I always have full sky maps:

```python
shape, wcs = pixell.enmap.fullsky_geometry(
    output_car_resol.to_value(u.radian), dims=(3,)
)   
ainfo = pixell.sharp.alm_info(lmax=lmax)
output_maps.append(
    pixell.curvedsky.alm2map(
        alm, pixell.enmap.empty(shape, wcs), ainfo=ainfo
    )   
    * unit
)
```

How do I modify the code to make it output the maps in the Standard SO CAR geometry?

**Comment by AdriJD** [comment]
Hi @zonca, for full sky maps we have to decide where the reference Y pixel sits, I think putting the reference pixel at RA, DEC = (0,0) is the most natural. The following implementation would be most in line with the above script:

```
def _get_fullsky_car_fejer1_geometry(res):
    """ Get map shape and WCS for full sky Fejer1 CAR variant with reference
    pixel at ra, dec = 0, 0.                                                                                                                                                                            
                                                                                                                                                                                                                                          
    Parameters                                                                                                                                                                                                                            
    ----------                                                                                                                                                                                                                            
    res : float                                                                                                                                                                                                                           
        CAR resolution (same for X and Y) in radians.                                                                                                                                                                                         
                                                                                                                                                                                                                                          
    Returns                                                                                                                                                                                                                               
    -------                                                                                                                                                                                                                               
    shape : tuple of int                                                                                                                                                                                                                  
        Shape of CAR geometry.                                                                                                                                                                                                            
    wcs : astropy.WCS                                                                                                                                                                                                                     
        WCS object describing CAR geometry.                                                                                                                                                                                               
    """

    # Create full sky geometry for Clenshaw–Curtis CAR variant.
    box = np.asarray([[-90.,  180.], [90., -180.]])
    shape, wcs = enmap.geometry(np.radians(box), res=res, proj='CAR')                                                                                                                                                                          
    # Convert to Fejer1 geometry.                                                                                                                                                                                                         
    wcs.wcs.crpix[1] -= 0.5
    shape = (shape[0] - 1, shape[1])
    return shape, wcs
```

This will result in a full sky Fejer1 CAR geometry with:
```
CDELT1 = -0.00833333
CDELT2 = 0.00833333
CRPIX1 = 21601.0
CRPIX2 = 10800.5
CRVAL1 = 0.0
CRVAL2 = 0.0
NAXIS1 = 43200
NAXIS2 = 21599
```

The resulting maps are compatible with the maps that Reijo will produce. Using `enmap.extract` people will be able to extract the patch corresponding to SO noise simulations from these full sky signal simulations without interpolation.
```
# Assume m_noise has been created by calling: python3 get_area_file.py --res 0.5 --proj car " -63:25,180:-180" area_file_00.5.fits
shape, wcs = _get_fullsky_car_fejer1_geometry(res)
m_signal = enmap.zeros(shape, wcs)
m_signal_in_noise_patch = enmap.extract(m_signal, m_noise.shape, m_noise.wcs)
```

**however** while testing the above I think I have found a bug in `pixell`. Consider the following:
```
res = np.radians(10 / 60)
shape, wcs = _get_fullsky_car_fejer1_geometry(res)
omap = enmap.zeros(shape, wcs)
lmax = (shape[0] - 1) // 2
ainfo = sharp.alm_info(lmax)
alm = curvedsky.rand_alm_white(ainfo)
alm[:ainfo.lmax + 1] = np.real(alm[:ainfo.lmax + 1])
oalm = curvedsky.map2alm(curvedsky.alm2map(alm, omap, ainfo=ainfo), alm.copy(), ainfo=ainfo)
np.allclose(alm, oalm)
```
which returns `False`. 

I'm checking with Sigurd what's going wrong. Hopefully get back to you soon with the final answer.

**Comment by zatkins2** [comment]
Not sure if related to your bug, but I think `NAXIS2` for a Fejer1 geometry with 0.5 arcmin pixels should be `21600`. E.g., the CAR geometry is always 1 more than you might expect (`5401` for 2 arcmin pixels, `10801` for 1 arcmin pixels, `21601` for 0.5 arcmin pixels), and Fejer1 is one less than that. Then the `lmax` for a Fejer1 geometry should be the same as for a CAR geometry I think (`5400` for 2 arcmin pixels, `10801` for 1 arcmin pixels, `21600` for 0.5 arcmin pixels), not one less than for CAR.

**Comment by AdriJD** [comment]
Hi Zach, good point that you would expect `NAXIS2` to be `21601` for the CC CAR variant. I think that is an issue with the call to `enmap.geometry`, i.e.:
```
box = np.asarray([[-90.,  180.], [90., -180.]])
shape, wcs = enmap.geometry(np.radians(box), res=res, proj='CAR')
```
If I change `box` to `box = np.asarray([[-90.,  180.], [90. + 0.5 / 60, -180.]])` you get the expected `21601` number of pixels.

It does not solve the problem though, I still think there is a bug. 

For example, if I do:
```
box = np.asarray([[-90.,  180.], [90. + 5 / 60, -180.]])
shape, wcs = enmap.geometry(np.radians(box), res=res, proj='CAR')                                             
wcs.wcs.crpix[1] -= 0.5
shape = (shape[0] - 1, shape[1])

minfo = curvedsky.get_minfo(shape, wcs, quad=True)
```
I find that `np.degrees(enmap.corners(shape, wcs, corner=False))` yields `[[ -89.95833333 -180. ] [  90.04166667 -540. ]]`, i.e. our most northern ring sits above the north pole. So this geometry should not work. However, the `minfo` object that `pixell` creates and passes to `libsharp` shows the correct fejer1 pixel locations: `print(np.degrees(minfo.theta - np.pi / 2))` gives `[ 89.95833333  89.875       89.79166667 ... -89.79166667 -89.875
 -89.95833333]` and the transforms become lossless. So there is something wrong with the conversion from enmaps to libsharp minfos.
 
About the lmax values: I believe the theory is as follows: `N + 1` Clenshaw-Curtis points are needed to integrate a polynomial of order `N`. However, only `N` Fejer1 points are needed for that polynomial, see e.g. slides 3 and 4 here: https://people.math.ethz.ch/~waldvoge/Papers/oxford.pdf . So strictly speaking I could have used `lmax = (shape[0]) // 2` instead of `lmax = (shape[0] - 1) // 2` (although in this case it gives the same value for `lmax`). The reason why there is a factor 2 is that the a spherical harmonic of order `lmax` is given by a polynomial of order `lmax` in `cos theta`. So you would think that you need just `lmax` points to perform `alm2map` in a lossless manner. However, then you forget that you are dealing with a product of the map and the spherical harmonic as integrand, i.e. `alm = int map * SH` and a map that is bandlimited at `lmax` is also given by a polynomial of maximum order `lmax`. So the integrand is a product of two polymials with combined maximum order of `2 lmax`. The Clenshaw Curtis and Fejer quadrature rules still behave well when you violate this rule (which is why we often get away with that) but you should not expect lossless transforms anymore, see this paper for example: https://www.math.stonybrook.edu/~bishop/classes/math331.F17/Projects/GaussVsCC.pdf

**Comment by zatkins2** [comment]
Weird, yes, sounds like the bug is indeed between enmap and libsharp! Agreed on the `lmax` factor of 2, I should have been more clear: I was imagining that `21599//2` would give `10799`, which might lose information in the transform. But yes I agree that with the right number of y pixels, whether or not one does `(shape[0] - 1)//2` or `shape[0]//2`will no longer matter.

**Comment by AdriJD** [comment]
I talked to Sigurd and he pushed some changes to `pixell`, see this commit: https://github.com/simonsobs/pixell/commits/a82fdf5e0ed02e4f040849ef9a68e10370dd8fd8 

The summary is that `enmap.fullsky_geometry` now has an option to produce fullsky `fejer1` CAR maps. There was a bug in `enmap.corners` that caused most of the confusion in the above posts. This has also been fixed in the above commit.

@zonca you can just update this call in `mapsims/mapsims/runner.py`
```
car_shape, car_wcs = pixell.enmap.fullsky_geometry(res=car_resolution.to_value(u.radian))
```
to 
```
car_shape, car_wcs = pixell.enmap.fullsky_geometry(res=car_resolution.to_value(u.radian), variant="fejer1")
```
The resulting geometry has:
```
CDELT1 = -0.00833333
CDELT2 = 0.00833333
CRPIX1 = 21600.5
CRPIX2 = 10800.5
CRVAL1 = 0.004167
CRVAL2 = 0.0
NAXIS1 = 43200
NAXIS2 = 21600
```
and is still compatible with the cut-sky maps that Reijo would produce through the use of `enmap.extract`. We have checked that the spherical harmonic transforms are now lossless.

**Comment by zonca** [comment]
Very good, thanks!

**Comment by zonca** [comment]
ok, this is available in pixell 0.17.3 and implemented in `mapsims`
